Xdmod Application Information Document (AID)

rob-baron opened the pull request:
Been asked to write up my thoughts and answer questions on xdmod.  Let me know if you have any questions that I should to answer.

rob-baron committed:
Start of Life for this document

This will be the last PR that I will submit for this project
as I have been asked to only document the status of this project
before I leave on 06-SEP-2023

Feel free to comment on it for parts that you would like more
explain on, It is still a work in progress and will be until
I officially leave.

rob-baron committed:
Just some minor edits

Finished one thought and added a section

rob-baron committed:
Added more details

naved001 commented on XdmodAID.md:
```diff
@@ -0,0 +1,265 @@
+Most of the time it runs withing a couple of minutes, sometimes it takes longer
+and will ocassionally fail due to a time out when it cannot lock certain tables.
+
+#xdmod-openstack
```
did you mean to put anything in this section?

rob-baron replied:
I guess it is not obvious to the order that I am doing things.

I laid out sections that I want to eventually fill out.  The xdmod-openstack section is a bit larger and I started on it today.

rob-baron replied:
Im in the middle of the xdmod-openstack section - a lot more to write

naved001 commented on XdmodAID.md:
````diff
@@ -0,0 +1,265 @@
+
+The structure is as follows:
+```
+```
````
the structure is missing here.

naved001 commented on XdmodAID.md:
```diff
@@ -0,0 +1,265 @@
+    group.csv      - has the mapping of nerc project to PIs
+    names.csv      - has the mapping of openshift namespaces and openstack projects to
+                     the nerc project
+    pi2project.csv - maps the openstack project to the openstack project
```
openstack project to *nerc* project?

rob-baron replied:
The pi2project is needed for the cloud hierarchy.  The way that the xdmod team uses this is to map the openstack project to the name in the names file.  I'll put in an example.

Sparing the confusing details that I could put in, pi2projects.csv is in essence mapping to itself.

rob-baron replied:
I have reworked this section, hopefully it is a bit clearer for you.

naved001 commented on XdmodAID.md:
```diff
@@ -0,0 +1,265 @@
+to find the dump process had finished.  The bigger issue was copying the data off as and unexpected EOF occurred
+several times later that evening.  In the morning, running the same command worked without issue.
+
+#Getting data out
```
Looks like this section is missing

rob-baron replied:
This section has been added

naved001 reviewed (commented):
thank you for the write up @rob-baron

rob-baron committed:
Added more details.

rob-baron committed:
More additions.

rob-baron committed:
gave a more detailed explation of the xdmod-hierarchy

This was to address naved's understandable confusion - I was
obviously too terse in my explaintion.

rob-baron committed:
More sections added

rob-baron committed:
A minor addition

rob-baron committed:
still more additions

rob-baron committed:
some more additions

rob-baron committed:
Added more to xdmod-openstack

rob-baron committed:
A bit of reoganization

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,853 @@
+
+As it is currently being used for billing and showback on a HPC cluster
+at Harvard, and it initally seemed to be a turnkey type system, we started
+to set this it up on an openshift cluster.  Unfortunately, we have found that it
```
remove up

rob-baron replied:
I had to read this twice to figure out what you meant.  I decided that you mean "remove 'up'" as opposed to removing more.

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,853 @@
+is not exactly a turn key system.
+
+Xdmod is currently designed to work on a stand alone
+server.  The do use docker to develop and to test it. Although to distruite it,
```
The -> They

joachimweyl replied:
distruite -> distribute

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,853 @@
+On the xdmod side, there is no support for OpenShift, and the OpenStack
+support is for an old version of OpenStack.
+
+Addtionally, there the environment the environment that we are deploying
```
remove there

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,853 @@
+#Introduction
```
after all # in the doc

rob-baron replied:
done.

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,853 @@
+could be run from ephemeral storage.
+
+From my perspective, this lack of performance on writes will make the cluster
+unusable for a wider array of applications and will a distinct pain point for
```
will a -> will be a

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,853 @@
+data flows from cloud_openstack and/or cloud_generic into the cloud_common
+
+The database field sizes are increased using the definitions within the json
+files as opposed using sql to alter the table definition.
```
using -> to using

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,853 @@
+##Transaction logs
+
+One thing that I have tried to find and been unsuccessful so far is to get
+transaction logs.  The closest thing that I have hear of is the OpenShift
```
hear of -> found

joachimweyl commented on XdmodAID.md:
````diff
@@ -0,0 +1,853 @@
+##disavantages to using xdmod
+```
+  1. Not turnkey for cloud services
+  2. No built in support for openshift
````
worth running a replace on openstack and openshift to change them to OpenStack & OpenShift to keep things the same.

rob-baron replied:
I've done this, but just to clarify, there is the command "xdmod-openstack" which is all lower case and the separate word "OpenStack" - likewise for OpenShift

joachimweyl replied:
good point, well that's too bad, I am sure you found a smart way to do it like with regex :)

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,853 @@
+support as a priority as they themselves don't deploy any current cloud
+service.
+
+It would not suprise me if we decide to drop xdmod only to develop
```
only to develop what?

joachimweyl replied:
this still needs attention.

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,853 @@
+#Dockerfile.moc-xdmod
+Initailly I copied over the docker files from the xdmod github repository as I
+figured there was the distinct possiblity that some restructuring would be
+necessary.  However, I also wanted to see if the original docker files could be used
```
remove However,

joachimweyl replied:
thoughts?

rob-baron replied:
removed 'However'

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,853 @@
+Initailly I copied over the docker files from the xdmod github repository as I
+figured there was the distinct possiblity that some restructuring would be
+necessary.  However, I also wanted to see if the original docker files could be used
+as is, as this would potentally make upgrading easier as we would be using what
```
as this -> this

joachimweyl replied:
this still needs attention.

rob-baron replied:
Just pushed this - should appear soon.

rob-baron committed:
Added ' ' between '#' and headings.

Added some more explaination as did a quick look over.

joachimweyl commented on XdmodAID.md:
````diff
@@ -0,0 +1,853 @@
+    supremm - should give some performance metrics for openshift
+```
+A resource can be a part of the realms, as I haven't really started to work on the
+storage resources (possibly part of a storage realm), I have confirmed this yet
````
this sentence needs a look, I am not sure if you meant to say that you have not confirmed this yet or that you have. "I have confirmed this by actually setting them up." or "I have not confirmed this yet by actually setting them up."

joachimweyl replied:
this still needs attention.

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+Alernatively, xdmod-init or xdmod-setup can be run from the command line if
+one uses rsh to gain access to the xdmod-ui container.  This often requires
+the additional step of restarting the xdmod-ui pod as the many changes require
+the web process to be retarted.
```
retarted -> restarted

joachimweyl replied:
this still needs attention.

rob-baron replied:
done.

rob-baron committed:
Some edits requested by kim

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+and we do not currently support RWX volumes, this needs to be run in parallel
+to the UI.
+
+Most of the time it runs withing a couple of minutes, sometimes it takes longer
```
withing -> within

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+
+In order to get the information about VMs, we went with the nova API.  Since the
+Nova API only gives information on VMs that are not deleted, by collectings data
+every 5 minutes, we expected to to catch when VMs topped running/were deleted.
```
to to -> to
topped -> stopped

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+
+In both the "OpenShift" and the "Generic" structure there is a way to
+track mounted volumes to the VM.  From an initial glance, something like
+this should be done to track volume useage, doing so here will not track
```
will not -> will

rob-baron replied:
The next line for context is "volumes that are allocated, but not mounted."

I guess I need to include the data structure, however, it assumes that any volume is attached.  There is no provision made for unattached volumes.  This is why implementing storage metrics is important.

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+for cloud data.  I was also able to use their test data to understand the details of
+the format.
+
+There seem to be 2 stages of verification.  The first is to confirm that file has
```
file -> the file

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+There seem to be 2 stages of verification.  The first is to confirm that file has
+the correct format.  Generally, the file should have strings.  The second stage
+will unpack the data from the first stage and ensure that the second stage has
+the correct format before it is inserted into database.  There were several
```
into -> into the

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+fields that required an integer inside of a string.
+
+Despite the xdmod team saying that it doesn't matter what a particular field is,
+what they mean is that as long as it can pass validaction, it may not be currently
```
validaction -> the validity check
or maybe verification?

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+
+The flavor_id is actually tracked in xdmod, and is required to be an integer
+within a string or it will not pass the validation.  The version of OpenStack
+that they were using has it as a integer OpenStack.  In the current version
```
a -> an
remove OpenStack from this line

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+
+The error messages that xdmod generates via the package they use for validation
+are confusing at best.  The error message will tell you that some type was expected
+where a different type was found.  Generally, I created files that 1 record
```
that -> with

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+
+## Some specific events that needed to be translated from current to past
+
+Becase xdmod suppored an older version of OpenStack, the many of the events
```
Becase -> Because

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+
+In order to report to insititutions, departments and PIs; a system wide hierarchy can
+be defined to aggreate statistics at the appropriate level.  This is done using
+csv files that define each level - the xdmod documentation  explaination can be
```
`__explaination` -> `_explanation`

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+
+Depending on if it is a HPC job or a cloud resource there are different levels to
+the hierarchy.  This is due to jobs not having a project name associated with
+them.  To side by side compare:
```
decide if you want one or two spaces after periods and then make sure they are all the same.

rob-baron replied:
ok, 2 spaces after a single '.'

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+Defining the first 3 levels of the hierarchy is done via the hierarchy.json file
+just as explained in the xdmod documentation.
+
+Unfortunately, when developing this used different names to define the PI
```
used -> xdmod used
or maybe instead of xdmod, I? I am not sure of the desired info here.

joachimweyl commented on XdmodAID.md:
````diff
@@ -0,0 +1,868 @@
+                    OpenStack/OpenShift project
+```
+
+What was implemented was a basic history table.  This was done in a simplistic way possible,
````
history -> hierarchy
in a simplistic way -> in as simplistic a way as

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+What was implemented was a basic history table.  This was done in a simplistic way possible,
+while being mindful of the non-perofromant file system.
+
+I anticipate the need to recreate the hierarchy files some arbitary date.  Although this feature
```
files -> files at

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+
+The obvious way to speed this up is to create a table that just has the most recent
+active entires. With such a table, you don't need to do any processing after selecting
+from it, and it is just a select, so this will be a fast operation giving our backend
```
giving -> given

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+add this in the future would require performance testing to ensure that SQLAlchemy is
+doing what we think it is doing.
+
+It was also suggested that I use a "limit by 1" since all I'm interested in just the most
```
in -> in is

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+
+It was also suggested that I use a "limit by 1" since all I'm interested in just the most
+recent entry.  Although, on the surface this seems reasonable, there are some problems
+with it.  To pull of the most recent one will require a sorting (an order by) which creats
```
of -> from

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+It was also suggested that I use a "limit by 1" since all I'm interested in just the most
+recent entry.  Although, on the surface this seems reasonable, there are some problems
+with it.  To pull of the most recent one will require a sorting (an order by) which creats
+and writing a virtual table to perform the sort in and then to return the first record.
```
writing -> writes
in and -> , and

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+recent entry.  Although, on the surface this seems reasonable, there are some problems
+with it.  To pull of the most recent one will require a sorting (an order by) which creats
+and writing a virtual table to perform the sort in and then to return the first record.
+Furthermore, we would have to a query for each hierarchy_id in the hierarchy table.  I
```
to a -> to

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+with many universities.  Finally, I used my name at the PI level in the hierarchy
+file.
+
+The group.csv file relates cold front project names to the id of the PI in the hierarchy
```
cold front -> ColdFront

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+to an openstack project that is found in the names.csv.
+
+The examples found in the xdmod documentation use abbreviations for unique keys. This
+doesn't suite an automated system as there are cases where there are multiple
```
suite -> suit

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+at the time there were 5 westminster colleges.  Since, we really don't need a
+human readable ID, I found using numbers to be more convienent.
+
+After working throught that example, it is obvious that many of the hierarchy items
```
throught -> through

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+It was suggested that we use a cryptological signiture to give unique ids as the possiblity
+of a data collision is miniscule. The advantage here would be that there would be no
+database access to determine what the ID would be.  It turned out to be a bit more
+complicated, as each item in the hierarchy needed it's own unique.  It was simpler create
```
unique what?

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+settled on using it the way it was designed.
+
+Another option would be to use a service container to host mariadb, however,
+this initially seemed a bit more complicated than just
```
just what?

rob-baron replied:
I was probably falling asleep when I wrote that and dreamed that I had finished my thought.

I hope what I wrote is better.

joachimweyl commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+Usually less than an hour.  When I did this the last time, I kiecked off the
+mysqldump command and after creating the pod to pull the PV data, came back to
+find the dump process had finished.  The bigger issue was copying the data off
+as and unexpected EOF occurred several times later that evening.  In the
```
and -> an

joachimweyl commented on XdmodAID.md:
````diff
@@ -0,0 +1,868 @@
+# storgae reporting
+
+There is a section of xdmod used for reporting on storage.  It is covered here:
+```
````
assume this is on its way.

rob-baron committed:
some edits requested by kim

And some addtions implied.

rob-baron commented on XdmodAID.md:
```diff
@@ -0,0 +1,868 @@
+
+This would also work with how I deploy xdmod on openshift using build
+configurations to test before actual deployment.  Furthermore, it could be used
+to test the production system insitu.
```
in situ

rob-baron committed:
More edits suggested/implied by kim

rob-baron committed:
edits suggested/implied by kim.

rob-baron committed:
Made some updates

rob-baron committed:
some more additions

rob-baron committed:
some more edits

rob-baron committed:
Merge branch 'main' of https://github.com/rob-baron/xdmod-cntr into xdmod-AID

rob-baron committed:
removed 'However'

rob-baron committed:
corrected spelling of 'restarted'

rob-baron committed:
fixed some confusing wording about realms and resources

rob-baron committed:
fixed some predictions.

rob-baron committed:
corrected some confusing wording.

rob-baron committed:
made some more corrections

rob-baron committed:
hopefully the last commit :)
